# Guns in USA

https://www.atf.gov/resource-center/data-statistics

In [121]:
import pandas as pd
import numpy as np
import sidetable
import matplotlib.pyplot as plt
import seaborn as sns

<font color= "cian"> data from: https://fivethirtyeight.com/features/gun-deaths-about/ 

<font color= "cian"> Loading the data

In [122]:
df = pd.read_csv("full_data.csv", index_col = 0)

<font color= "cian"> Exploring the data: I want to explore the shape of our dataframe(df), unique values, typo errors, types of data, columns...

In [123]:
df.head()

year  month   intent  police sex   age                    race  hispanic  \
1  2012      1  Suicide       0   M  34.0  Asian/Pacific Islander       100   
2  2012      1  Suicide       0   F  21.0                   White       100   
3  2012      1  Suicide       0   M  60.0                   White       100   
4  2012      2  Suicide       0   M  64.0                   White       100   
5  2012      2  Suicide       0   M  31.0                   White       100   

             place     education  
1             Home           BA+  
2           Street  Some college  
3  Other specified           BA+  
4             Home           BA+  
5  Other specified        HS/GED

In [124]:
df.shape

(100798, 10)

In [125]:
df.dtypes

year           int64
month          int64
intent        object
police         int64
sex           object
age          float64
race          object
hispanic       int64
place         object
education     object
dtype: object

<font color= "cian"> All the types look correctly.

<font color= "cian"> Exploring nulls

In [126]:
percent_missing = pd.DataFrame(df.isna().sum()*100 / len(df))
percent_missing

0
year       0.000000
month      0.000000
intent     0.000992
police     0.000000
sex        0.000000
age        0.017857
race       0.000000
hispanic   0.000000
place      1.373043
education  1.410742

<font color= "cian">

<font color= "cian"> I assume that this 1% of nulls is not so much, then I will eliminate them.

In [127]:
df.dropna(inplace= True)

In [128]:
# Check if there is nans after clean them
percent_missing = pd.DataFrame(df.isna().sum()*100 / len(df))
percent_missing

0
year       0.0
month      0.0
intent     0.0
police     0.0
sex        0.0
age        0.0
race       0.0
hispanic   0.0
place      0.0
education  0.0

<font color= "cian"> Are there duplicates?

In [129]:
df.duplicated().sum()*100/len(df)

39.190940162220066

<font color= "cian"> We have many duplicates but is logical because of the type our data. I decide to maintain them. 

<font color= "cian"> The columns of our dataset

In [130]:
df.columns

Index(['year', 'month', 'intent', 'police', 'sex', 'age', 'race', 'hispanic',
       'place', 'education'],
      dtype='object')

<font color= "cian"> Different categories or types of gun deaths (unique values)

In [131]:
for col in df.columns:
    print("\n In the variable ----------> ", col,"----- > These are the unique values:  ", df[col].unique())


 In the variable ---------->  year ----- > These are the unique values:   [2012 2013 2014]

 In the variable ---------->  month ----- > These are the unique values:   [ 1  2  3  4  5  6  7  8  9 10 11 12]

 In the variable ---------->  intent ----- > These are the unique values:   ['Suicide' 'Undetermined' 'Accidental' 'Homicide']

 In the variable ---------->  police ----- > These are the unique values:   [0 1]

 In the variable ---------->  sex ----- > These are the unique values:   ['M' 'F']

 In the variable ---------->  age ----- > These are the unique values:   [ 34.  21.  60.  64.  31.  17.  48.  41.  50.  30.  43.  27.  55.  53.
  51.  65.  52.  47.  19.  22.  56.  38.  25.  26.  23.  87.  28.  39.
  59.  61.  12.  40.  57.  20.  62.  71.  44.  29.  24.   9.  77.  18.
  36.  49.  13.  33.  90.  63.  46.  32.  84.  68.  89.  58.  78.  66.
  81.  54.  45.  37.  35.  42.  67.  69.  72.  82.  75.  73.  15.  83.
  74.  86.  16.  88.  80.  79.  91.  76.  14.  95.  96.  70.   7.  85.

<font color= "red"> Investigate what is each column and their values. Hispanic?

In [132]:
df.describe()

year         month        police           age      hispanic
count  98015.000000  98015.000000  98015.000000  98015.000000  98015.000000
mean    2013.000561      6.567780      0.000194     43.900209    112.420650
std        0.816136      3.406984      0.013922     19.522908     49.320698
min     2012.000000      1.000000      0.000000      0.000000    100.000000
25%     2012.000000      4.000000      0.000000     27.000000    100.000000
50%     2013.000000      7.000000      0.000000     42.000000    100.000000
75%     2014.000000      9.000000      0.000000     58.000000    100.000000
max     2014.000000     12.000000      1.000000    107.000000    998.000000

<font color= "cian"> We have to investigate what each variable is 

In [133]:
df.describe(include="object")

intent    sex   race  place education
count     98015  98015  98015  98015     98015
unique        4      2      5     10         4
top     Suicide      M  White   Home    HS/GED
freq      62291  83835  64598  59622     42258

<font color= "cian"> I explore outliers

In [134]:
# I select only the numerical variables
df_num = df.select_dtypes(include = np.number)
df_num.head()

year  month  police   age  hispanic
1  2012      1       0  34.0       100
2  2012      1       0  21.0       100
3  2012      1       0  60.0       100
4  2012      2       0  64.0       100
5  2012      2       0  31.0       100

In [135]:
# Select just the categorical variables
df_categ = df.select_dtypes(include= "object")
df_categ.head()

intent sex                    race            place     education
1  Suicide   M  Asian/Pacific Islander             Home           BA+
2  Suicide   F                   White           Street  Some college
3  Suicide   M                   White  Other specified           BA+
4  Suicide   M                   White             Home           BA+
5  Suicide   M                   White  Other specified        HS/GED

In [137]:
df_num.columns

Index(['year', 'month', 'police', 'age', 'hispanic'], dtype='object')

In [138]:
# Calculate IQR
q3 = np.percentile(df_num["age"],75)
q3

58.0

In [139]:
q1 = np.percentile(df_num["age"],25)
q1

27.0

In [140]:
iqr = q3 - q1
iqr

31.0

In [141]:
# Calculate the whiskers of the box plot
outliers_sep = iqr * 1.5
outliers_sep

46.5

In [142]:
outliers_data = df_num[(df_num["age"] < q1 - outliers_sep) | (df["age"] > q3 + outliers_sep)]
outliers_data

year  month  police    age  hispanic
34903  2013      9       0  107.0       100

<font color= "cian"> There is one outlier in `age` variable with the value 107. I decide to delete it.

In [175]:
outlier_list = list(outliers_data.index)
outlier_list


[34903]

In [150]:
df_clean = df_num.drop(outliers_data.index)
df_clean.shape

(98014, 5)

In [151]:
df_num.shape

(98015, 5)

<font color= "cian"> Here we see comparing `df_num` with `df_clean` and we delete one row. Just how we want.

In [ ]:
df_num.iloc[[34903]]

In [168]:
df_clean.iloc[[34903]]

year  month  police   age  hispanic
35909  2013     11       0  78.0       100

In [ ]:
# create a function to check the outliers 

_________________________________________

<font color= "cian"> How many gun deaths are there by race?

In [144]:
df["race"].value_counts()

White                             64598
Black                             22675
Hispanic                           8603
Asian/Pacific Islander             1261
Native American/Native Alaskan      878
Name: race, dtype: int64

In [145]:
df.stb.freq(["race"])

race  count    percent  cumulative_count  \
0                           White  64598  65.906239             64598   
1                           Black  22675  23.134214             87273   
2                        Hispanic   8603   8.777228             95876   
3          Asian/Pacific Islander   1261   1.286538             97137   
4  Native American/Native Alaskan    878   0.895781             98015   

   cumulative_percent  
0           65.906239  
1           89.040453  
2           97.817681  
3           99.104219  
4          100.000000

<font color= "cian">  Mean age by race

In [146]:
mean_race = round(df.groupby(["race"])["age"].mean(),2)
mean_race

race
Asian/Pacific Islander            38.61
Black                             31.06
Hispanic                          33.14
Native American/Native Alaskan    36.49
White                             50.04
Name: age, dtype: float64

<font color= "cian"> Trying colors